pkg/cmd/project/link/link.go

```diff
@@ -0,0 +1,162 @@
+		Short: "Link a project to a repository or a team",
+		Use:   "link [<number>] [flag]",
+		Example: heredoc.Doc(`
+			# link monalisa's project 1 to her repository "my_repo"
+			gh project link 1 --owner monalisa --repo my_repo
+
+			# link monalisa's organization's project 1 to her team "my_team"
+			gh project link 1 --owner my_organization --team my_team
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := client.New(f)
+			if err != nil {
+				return err
+			}
+
+			if len(args) == 1 {
+				num, err := strconv.ParseInt(args[0], 10, 32)
+				if err != nil {
+					return cmdutil.FlagErrorf("invalid number: %v", args[0])
+				}
+				opts.number = int32(num)
+			}
+
+			config := linkConfig{
+				httpClient: f.HttpClient,
+				client:     client,
+				opts:       opts,
+				io:         f.IOStreams,
+			}
+
+			if err := cmdutil.MutuallyExclusive("specify only one of `--repo` or `--team`", opts.repo != "", opts.team != ""); err != nil {
+				return err
+			}
+			if err := cmdutil.MutuallyExclusive("specify either `--repo` or `--team`", opts.repo == "", opts.team == ""); err != nil {
+				return err
+			}
+
+			// allow testing of the command without actually running it
+			if runF != nil {
+				return runF(config)
+			}
+			return runLink(config)
+		},
+	}
+
+	linkCmd.Flags().StringVar(&opts.owner, "owner", "", "Login of the owner. Use \"@me\" for the current user.")
+	linkCmd.Flags().StringVarP(&opts.repo, "repo", "R", "", "The repository to be linked to this project")
+	linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be linked to this project")
+	cmdutil.AddFormatFlags(linkCmd, &opts.exporter)
+
+	return linkCmd
+}
+
+func runLink(config linkConfig) error {
+	canPrompt := config.io.CanPrompt()
+	owner, err := config.client.NewOwner(canPrompt, config.opts.owner)
+	if err != nil {
+		return err
+	}
+
+	project, err := config.client.NewProject(canPrompt, owner, config.opts.number, false)
+	if err != nil {
+		return err
+	}
+	config.opts.projectID = project.ID
+
+	httpClient, err := config.httpClient()
+	if err != nil {
+		return err
+	}
+	c := api.NewClientFromHTTP(httpClient)
+
+	if config.opts.repo != "" {
+		return linkRepo(c, owner, config)
+	} else if config.opts.team != "" {
+		return linkTeam(c, owner, config)
+	}
+	return nil
+}
+
+func linkRepo(c *api.Client, owner *queries.Owner, config linkConfig) error {
+	repo, err := api.GitHubRepo(c, ghrepo.New(owner.Login, config.opts.repo))
+	if err != nil {
+		return err
+	}
+	config.opts.repoID = repo.ID
+
+	result, err := config.client.LinkProjectToRepository(config.opts.projectID, config.opts.repoID)
+	if err != nil {
+		return err
+	}
+
+	if config.opts.exporter != nil {
+		return config.opts.exporter.Write(config.io, result)
+	}
+	return printResults(config, result.URL)
+}
+
+func linkTeam(c *api.Client, owner *queries.Owner, config linkConfig) error {
+	team, err := api.OrganizationTeam(c, ghrepo.New(owner.Login, ""), config.opts.team)
```

Comment:
I agree with you. My initial idea was to keep the parameters of functions in the same file as consistent as possible, even if it looks a bit odd. However, with the question you've raised, I think implementing it your way would be better

Reply:
Should the `host` in `linkRepo` function be retrieved in the same way?

Reply:
@benebsiny I think you added a question and then deleted it cause I can't see it anymore but actually I think it was a really good question. Indeed I think for `linkRepo` we also need to allow the opportunity for the repo to be on different hosts.

I also have a question, if I were in a repo say `cli/cli` and I ran `gh project link 1` would you expect it to work? What about `GH_REPO=cli/cli gh project link 1`?

The reason I'm asking this is because other commands that work with a repo support this and the approach to solving `linkRepo` above will vary based on this design decision.

Reply:
> I think you added a question and then deleted it cause I can't see it anymore...

Yes, I asked the question, and I later found the answer (the answer is same as yours), so I deleted that question.

> I also have a question, if I were in a repo say cli/cli and I ran gh project link 1 would you expect it to work?

I think it's really a good feature. I can implement it.

Reply:
Yeh so in that regard I think you should look at something like the `issue` command and the `EnableRepoOverride` function. Sorry I don't have time to provide more guidance right now but see how it is used for `issue view` (or any other commands)